File: 695-max-area-of-island/695-max-area-of-island.c
#define EXPLORED 0;

struct Island{
    int** grid;
    int gridSize;
    int* gridColSize;
};

int isLand(struct Island* island, int row, int col) {
    return row >= 0 && row < island->gridSize &&
        col >= 0  && col < island->gridColSize[row];
}

int isUnexploredLand(struct Island* island, int row, int col) {
    return isLand(island, row, col) && island->grid[row][col] != EXPLORED;
}

int explore(struct Island* island, int row, int col) {
    if (isUnexploredLand(island, row, col)) {
        island->grid[row][col] = EXPLORED;
        
        return 1 + explore(island, row-1, col) + explore(island, row, col-1) +
            explore(island, row+1, col) + explore(island, row, col+1);
    }
    return 0;
}

int maxAreaOfIsland(int** grid, int gridSize, int* gridColSize){
    struct Island island;
    island.grid = grid;
    island.gridSize = gridSize;
    island.gridColSize = gridColSize;
    
    int maxArea = 0;
    for(int row = 0; row < gridSize; row++) 
        for(int col = 0; col < gridColSize[row]; col++) {
           if (isLand(&island, row, col)) {
               int area = explore(&island, row, col);
               if(area > maxArea)
                   maxArea = area;
           }
        }
    
    return maxArea;
}

File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.EXPLORED = -1
        self.grid = grid
        
        maxArea = 0
                
        for row, rowCells in enumerate(grid):
            for col, cell in enumerate(rowCells):
                if cell == 1:
                    area = self.bfs(row, col)
                    
                    if area > maxArea:
                        maxArea = area
        
        return maxArea
                    
    def bfs(self, row, col):
        def inGrid(cell):
            row, col = cell
            return row >= 0 and row < len(self.grid) and col >= 0 and col < len(self.grid[row])
            
        def isUnexplored(cell):
            row, col = cell
            return not self.grid[row][col] == self.EXPLORED
        
        def isLand(cell):
            row, col = cell
            return self.grid[row][col] == 1
        
        area = 0
        directions = [(0, -1), (-1, 0), (0, 1), (1, 0)]
        stack = deque()
        stack.append((row, col))
        
        while len(stack):
            row, col = stack.pop()
            
            if isUnexplored((row, col)):            
                self.grid[row][col] = self.EXPLORED
                area += 1
                
                for stepRow, stepCol in directions:
                    neighbour = (row + stepRow,  col + stepCol)
                    if inGrid(neighbour) and isLand(neighbour) and isUnexplored(neighbour):
                        stack.append(neighbour)
        
        return area